### PR TITLE
[Glossary] Proposing system description terminology

### PR DESCRIPTION
**Pull request by fred-r**
Proposal to add concepts discussed in:
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack/issues/33

### CODE REVIEW

File: doc/Glossary.md [hunk 1]
```diff
@@ -131,3 +131,29 @@ Those terms are not forbidden but frowned upon because of their history, ambigui
...
+
+- <a name="binary-program"></a>**Binary program**
+A binary program can be:
+  - A bare-metal binary,
+  - A monolithic OS binary and all the services running on top of it,
+  - An OS that can install modules. Each module can be a binary on its own but still from an execution perspective as the OS is required for this service to run we consider it as a single binary in term of execution context. 
+  - A virtual machine or interpreter 
+- <a name="execution-context"></a>**Execution Context**
+An execution context identifies the “cradle” allowing to run a binary program on a given hardware platform.  
+  - A context of execution has a 1-to-1 association to an HW execution unit.
```
Comment (jkrech) [inline doc/Glossary.md:148]: an HW ... -> a HW

File: doc/Glossary.md [hunk 2]
```diff
@@ -131,3 +131,29 @@ Those terms are not forbidden but frowned upon because of their history, ambigui
...
+- <a name="binary-program"></a>**Binary program**
+A binary program can be:
+  - A bare-metal binary,
+  - A monolithic OS binary and all the services running on top of it,
+  - An OS that can install modules. Each module can be a binary on its own but still from an execution perspective as the OS is required for this service to run we consider it as a single binary in term of execution context. 
+  - A virtual machine or interpreter 
+- <a name="execution-context"></a>**Execution Context**
+An execution context identifies the “cradle” allowing to run a binary program on a given hardware platform.  
+  - A context of execution has a 1-to-1 association to an HW execution unit. 
+  - If we support virtual machines or interpreters then the virtual machine is associated to one execution context. The programs interpreted by the virtual machine (VM) are stored as data sections in memory, and considered as execution modules within the VM's execution context.
```
Comment (jkrech) [inline doc/Glossary.md:149]: associated to -> with

File: doc/Glossary.md [hunk 3]
```diff
@@ -131,3 +131,29 @@ Those terms are not forbidden but frowned upon because of their history, ambigui
...
+- <a name="execution-domain"></a>**Execution Domain**
+An execution domain is a container aggregating all the platform resources (internal and external memories, peripherals) that an execution context can use “exclusively”. For shared resources (available for several execution contexts), we introduce the concept of shared execution domain.
+- <a name="execution-module"></a>**Execution Module**
+An execution module is an entity that has no standalone execution possibility. It relies on a “backbone" binary providing system services. So, an execution module is always associated to an execution context.
```
Comment (jkrech) [inline doc/Glossary.md:153]: associated to -> with

File: doc/Glossary.md [hunk 4]
```diff
@@ -131,3 +131,29 @@ Those terms are not forbidden but frowned upon because of their history, ambigui
...
+- <a name="HW-exec-unit"></a>**Hardware Execution Unit**
+An HW execution unit is either a cluster (i.e. group of cores), or a core, or a TrustZone world within a core (virtualization). This is the hardware entity (virtual or physical) running a binary program.
```
Comment (jkrech) [inline doc/Glossary.md:155]: An -> A HW ...

Comment (fred-r) [reply doc/Glossary.md]: Agreed and fixed.

**Comment by fred-r** [comment]
Joachim's comments taken into account.

Please note that solution and application project are out of scope of this section.
I guess we will add it in a project management section later on ?

We may converge with a solution being a collection of application projects :-)

**Review by LMESTM — APPROVED** [review]

**Comment by fred-r** [comment]
any update about this PR ?

**Comment by ReinhardKeil** [comment]
@fred-r is this pull request still relevant?

**Comment by fred-r** [comment]
Well, I think so.
Probably it needs to be adjusted to the implementation artefacts like csolution and cproject ?

**Comment by ReinhardKeil** [comment]
Should we add a section "Glossary" to https://github.com/Open-CMSIS-Pack/cmsis-toolbox/blob/main/docs/README.md?

**Comment by ReinhardKeil** [comment]
Document is under https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/pull/63/files

**Review by jkrech — APPROVED** [review]
This is a good start. We can always refine.

**Pull request merged**